e2e/test/MessageReceiveE2ETests.cs

```diff
@@ -165,16 +335,127 @@ private async Task ReceiveSingleMessage(TestDeviceType type, Client.TransportTyp
+                await deviceClient.CloseAsync().ConfigureAwait(false);
+                await serviceClient.CloseAsync().ConfigureAwait(false);
+            }
+        }
+
+        private async Task ReceiveAbandonedMessage(TestDeviceType type, Client.TransportType transport)
+        {
+            TestDevice testDevice = await TestDevice.GetTestDeviceAsync(DevicePrefix, type).ConfigureAwait(false);
+            
+            Client.Message receivedMessage1, receivedMessage2;
+            using (DeviceClient deviceClient = testDevice.CreateDeviceClient(transport))
+            using (ServiceClient serviceClient = ServiceClient.CreateFromConnectionString(Configuration.IoTHub.ConnectionString))
+            {
+                await deviceClient.OpenAsync().ConfigureAwait(false);
+                await serviceClient.OpenAsync().ConfigureAwait(false);
+                await serviceClient.PurgeMessageQueueAsync(testDevice.Id).ConfigureAwait(false);
+
+                (Message msg1, string messageId1, string payload1, string p1Value) = ComposeC2DTestMessage();
+                await serviceClient.SendAsync(testDevice.Id, msg1).ConfigureAwait(false);
+                (Message msg2, string messageId2, string payload2, string p2Value) = ComposeC2DTestMessage();
+                await serviceClient.SendAsync(testDevice.Id, msg2).ConfigureAwait(false);
+
+                receivedMessage1 = await deviceClient.ReceiveAsync().ConfigureAwait(false);
+
+                /* With AbandonAsync ordering of message is not guaranteed due to Prefetch Count set 
+                 * on AMQP transport settings as there could be pending messages in local queue which 
+                 * are received before receiving abandoned message so we are not testing when the message 
+                 * is received but whether it is received */
+                await deviceClient.AbandonAsync(receivedMessage1).ConfigureAwait(false);
+                receivedMessage2 = await deviceClient.ReceiveAsync().ConfigureAwait(false);
+                Client.Message receivedMessage3 = await deviceClient.ReceiveAsync().ConfigureAwait(false);
+
+                /* There is msg2 still in the queue and msg1 due to AbandonAsync hence check for NULL */
+                Assert.IsNotNull(receivedMessage2);
+                Assert.IsNotNull(receivedMessage3);
+
+                var prop2 = receivedMessage2.Properties.Single();
+                var prop3 = receivedMessage3.Properties.Single();
 
+                if (p1Value == prop2.Value)
+                {
+                    VerifyMessage(messageId1, payload1, p1Value, receivedMessage2);
+                    await deviceClient.CompleteAsync(receivedMessage2).ConfigureAwait(false);
+                }
+                else if (p1Value == prop3.Value)
+                {
+                    VerifyMessage(messageId1, payload1, p1Value, receivedMessage3);
+                    await deviceClient.CompleteAsync(receivedMessage3).ConfigureAwait(false);
```

Comment:
Shouldn't we be completing both the messages?

Reply:
yes moved it after if else